### PR TITLE
fix(quantization): per-stream Marlin workspaces to prevent cross-stream CTA deadlock under PD-Multiplexing

### LINKED ISSUE #37904: [PDMux] scheduler event loop deadlocks in decode_stream.synchronize() when a mixed decode+split-prefill batch follows a >=4K-token prefill (hybrid GDN models)
Reproduction script: **`test/manual/quant/test_pdmux_mixed_deadlock_repro.py`** in PR #37905 (self-contained, single GPU, ~10 minutes; a gist mirror exists but the in-repo copy is canonical). Verified in both directions on A100-SXM4-40GB: with only the stale-field prerequisite applied the deadlock reproduces (exit 2, scheduler blocked in `decode_stream.synchronize()` per the embedded py-spy dump); with the full PR all 3 sensitizing rounds complete cleanly (exit 0). The trigger is timing-sensitive: a single b1 (one concurrent prefill) pass is expected only ~30-70% pre-fix, so the script runs three victim+b1+b2 rounds by default. **Environment note**: the validation environment additionally carried the two-line stale-field fix (`self.gpu_id`→`self.ps.gpu_id`, `self.tp_size`→`self.ps.tp_size`) that current main still needs for PDMux to start at all — that fix is included in #37905. Fix: #37905.

## Environment

- GPU: NVIDIA A100-SXM4-40GB (108 SM, CC 8.0), driver 580.82.07
- SGLang: commit `1cf2b8c54d81802abc15dcf23a29b9cc687bc01e` (+ two-line stale-field
  fix in `multiplexing_mixin.py`: `self.gpu_id` → `self.ps.gpu_id`,
  `self.tp_size` → `self.ps.tp_size`; both fields no longer exist on
  `Scheduler` at this commit, so PDMux cannot start without it — separate bug)
- Model: `Qwen/Qwen3.8-27B-FP8` revision `017b9c7` (hybrid: 48×GatedDeltaNet +
  16×full attention), FP8-Marlin weight-only on SM80
- torch 2.13.0+cu129, `sglang-kernel 0.4.6.post1+cu129`
- Server flags: `--enable-pdmux --sm-group-num 3` with
  `manual_divisions: [[72, 36, 1]]`, `split_forward_token_budget: 256`,
  `--chunked-prefill-size -1 --disable-radix-cache --disable-cuda-graph
  --disable-overlap-schedule`

## Symptom

After the server completes one request with a **≥4K-token prompt** through the
split-prefill path, the next time the scheduler runs a decode batch and a
split-prefill batch concurrently, the event loop blocks forever inside:

```text
synchronize (torch/cuda/streams.py:108)
event_loop_pdmux (sglang/srt/multiplex/multiplexing_mixin.py:205)   # decode_stream.synchronize()
```

py-spy shows the loop's MainThread blocked at that synchronize; the decode
forward enqueued on the decode stream never completes. The watchdog thread
stays idle and never fires. The engine accepts no further work until the
process is killed. Requests that arrive after the hang never log another
"Prefill batch"/"Decode batch" line — the loop is stuck, not skipping work.

Single-request flows stay healthy after the trigger request: a subsequent
lone request (prefill + its own decode) completes normally. Only the
concurrent (mixed) pattern deadlocks.

## Reproduction

`upstream/repro_pdmux_mixed_deadlock.py` automates the sequence on any
A100/H100 with a qwen3_5-family checkpoint:

1. start server as above, wait ready
2. request A: 1024-token prompt, 32 output tokens (completes)
3. request B: **4096-token prompt**, 512 output tokens (completes)
4. request C: concurrent — 256-token decode plus a 1024-token prompt with 32
   output tokens, staggered by 0.25 s (the PDMux mixed pattern)

Step 4 times out (300 s cap) in 4/4 attempts; with step 2 removed or replaced
by a 1K-prompt request, step 4 completes in 6/6 attempts. A 1K-prompt request
with a long (512-token) decode does **not** trigger it, and neither
`/flush_cache` nor `/abort_request` is required — abort merely correlated in
an earlier test ordering.

## Characterization so far

- budget-independent: reproduces with `split_forward_token_budget` 256 and
  65536
- timing-sensitive: adding per-call CUDA event/stream API logging (file I/O)
  makes the deadlock disappear; an in-memory double-buffered tracer with a
  200 ms background flush preserves it
- CPU-side torch event accounting is clean at the hang: a low-overhead
  in-memory tracer (hooking `Event.record/wait/query`,
  `Stream.wait_stream/synchronize`) shows **no** `Event.wait` whose event was
  never re-recorded — the unsatisfied dependency is **below the torch API**,
  i.e. device-side, inside the decode forward's kernel chain
- captured interleaving right before the hang (tracer window, all times s):
  - `547.9986 ev.record Event@d6aa0 on Stream@d3a10` — the prior 4K request's
    final slice records `prefill_exe_done`; `ev.query -> False ×3 -> True`;
    merge + stream-group switch (`st.sync` of both new-group streams at
    `548.0030`)
  - `548.1041 → 548.3930`: the mixed phase cycles normally at ~95 ms per
    iteration, each iteration ending in `st.sync ExternalStream@373c0 done`
    (the new group's decode stream). **No torch-level event ops occur during
    the mixed phase**
  - `548.3930`: the next `decode_stream.synchronize()` (line 205) never
    returns; no further traced call happens for the remaining 300 s
- conclusion: the decode forward's kernel chain, enqueued on the group's
  decode stream, waits device-side on work ordered on the group's prefill
  stream (the concurrently-running split-prefill slice) — a circular
  device-side wait between the two green-context streams. Since the CPU
  thread never issues a torch-level wait for it, the dependency is created
  below torch (sgl_kernel / Triton runtime, or device-side flags in the
  hybrid linear-attention kernels)

## Root cause

`cuda-gdb --batch -p <scheduler> -ex "info cuda kernels"` taken during the
hang names the single active kernel:

```
Kernel  Dev Grid   Status  SMs Mask                          GridDim    BlockDim  Invocation
*    0   -   0  ...     Active  0x0000000000000000000000088000  (108,1,1) (256,1,1)
    sglang::device::marlin::Marlin<__nv_bfloat16, ...>()
```

`nvidia-smi pmon` during the hang shows the scheduler process at 99% SM —
the Marlin GEMM is **spinning**, not waiting on a stream event. Grid
`(108,1,1)` is the full-device SM count (`marlin_make_workspace` sizes its
cross-CTA workspace by `torch.cuda.get_device_properties(...).
multi_processor_count`), while the executing stream is a 36-SM
green-context stream.

The mechanism: every Marlin-using layer holds **one** workspace tensor
(`layer.workspace = marlin_make_workspace(device)`), and the kernel's
cross-CTA workspace protocol is only safe when all launches touching that
workspace are serialized on a single stream. PDMux breaks that assumption:
the decode forward (decode stream) and the split-prefill slice (prefill
stream) of the same layer execute **concurrently on two green-context
streams against the same per-layer workspace**; their CTAs interleave
workspace slots and one launch spins forever. This explains every observed
property of the bug: it needs a concurrent decode+split-prefill batch
(single-stream launches are serialized), it is independent of
abort/flush/decoding length, and the >=4K-prompt prior raises the overlap
probability of same-layer Marlin launches to near-certainty (long
multi-slice prefill sweeps every layer while decode steps run).

Separate (minor) gap found while reading: `Scheduler.abort_request` only
walks `waiting_queue`, `running_batch` and `last_batch`
(`inflight_batches = [self.running_batch, self.last_batch]`); requests
currently inside `split_prefill_batch` are invisible to `/abort_request`.

## Proposed fix

Give each executing stream its own Marlin workspace when PDMux stream
groups exist; nothing changes on non-PDMux paths:

- `marlin_utils.py`: add `get_marlin_workspace_for_forward(layer)` — returns
  `layer.workspace` unless PDMux stream groups exist, in which case return a
  per-`cuda.Stream` workspace from a lazily populated per-layer dict
  (`marlin_make_workspace` is a few KB, so per-stream copies are negligible).
- `marlin_utils_fp8.py::prepare_fp8_layer_for_marlin`: initialize
  `layer.marlin_stream_workspaces = {}` when stream groups exist.
- `fp8.py::Fp8LinearMethod.apply`: pass
  `workspace=get_marlin_workspace_for_forward(layer)`.

The same pattern should be applied to the other Marlin paths
(`marlin_utils_fp4.py`, GPTQ/AWQ Marlin, `fused_marlin_moe`) since they share
the workspace assumption. A concrete proposal patch against `1cf2b8c5` is
available in the project repository
(`third_party/marlin_workspace_pdmux_fix.proposal.patch`); for the PR we
will rebase it onto main and add a regression test. One caveat from our
validation: a naive two-stream test (submit one GEMM per stream, then
synchronize) does **not** reproduce the hang even with a shared workspace —
the barrier counters self-heal once the interfering queue drains. The test
must queue a whole forward's worth of launches per stream (e.g. 24+ per
stream) before synchronizing, mirroring how the scheduler stacks a decode
forward against a prefill slice; that pattern is what our standalone probe
used to bound the behavior.

## Performance impact of the fix

The fix changes only which pre-allocated workspace tensor is passed to the
kernel — no kernel, grid, or memory-layout change:

- steady-state hot-path cost: one `getattr` plus one dict lookup per linear
  apply (nanoseconds, once per layer per forward)
- lazy first-use allocation per (layer, stream): a 108-entry `torch.zeros`
  (432 B) memset on the executing stream, at most `#layers x #streams`
  times per process (~220 KB total for a 64-layer model with 6 green-context
  streams)
- non-PDMux paths are byte-for-byte unchanged (the per-stream map is only
  created when PDMux stream groups exist)

We measured the fix on a single A100-SXM4-40GB (Qwen3.8-27B-FP8, 64 layers,
48 GDN + 16 full-attention) with the PDMux mixed-batch suite:

- before the fix, the mixed decode+prefill cases deadlocked 3/3 (and, in a
  six-iteration A/B loop on the unfixed build, the first mixed case hung
  4/6 and the second 6/6 — the race is probabilistic per iteration);
- with the fix, the same cases completed on every attempt across two fast
  gates and the full correctness matrix (five overlap cases: 1k/4k/16k
  prefills at batch 1/2/8, each compared token-hash against a
  no-prefill-interference reference — all equal, zero gate failures);
- per-case wall times with the fix are within ~5-10% of healthy unfixed
  runs (mixed case 25.9-27.5 s vs 23.5-24.7 s healthy baseline; the delta
  is dominated by run-to-run variance on preemptible VMs, not by the fix —
  the hot path adds only a `getattr` + dict lookup per linear apply);
- decode `gen throughput` during mixed windows: 34-160 token/s depending
  on running batch size; prefill `input throughput` ~120 token/s while
  decode proceeds concurrently — i.e., the overlap PDMux exists to provide
  is intact with the fix.

A dedicated PDMux SLO-goodput comparison is separate follow-up work; this
fix claims no throughput improvement, only elimination of the deadlock.

## What we can provide

- deterministic-ish repro script and exact server commands
- py-spy dumps of the blocked loop (attached in the issue)
- full run artifacts (server logs with per-request scheduler lines)
- a Colab-safe reproduction session if needed

### PR DESCRIPTION
**Pull request by mikewong23571**
Fixes #37904

## Part 1 — prerequisite: repair stale Scheduler fields (PDMux cannot start on current main)

`multiplexing_mixin.py` still reads `self.gpu_id` / `self.tp_size`, fields that no longer exist on `Scheduler` (they moved to `ParallelState`). PDMux raises in `initialize_stream_groups` before the model loads. Replaced with `self.ps.gpu_id` / `self.ps.tp_size`, matching the pattern used elsewhere in `scheduler.py`.

## Part 2 — the deadlock fix

Under `--enable-pdmux`, the decode forward and the split-prefill slice of the **same layer** run concurrently on two green-context streams. Marlin kernels use `workspace` as a cross-CTA lock array (`barrier_acquire` waits for exact counter equality) and assume every launch touching one workspace is serialized on a single stream. Sharing one per-layer workspace lets both launches corrupt the barrier counters and a CTA spins forever — the scheduler hangs in `decode_stream.synchronize()` at 99% SM, and the watchdog never fires. Repro and device-side evidence (cuda-gdb naming the spinning Marlin kernel, py-spy stacks, measured before/after): #37904.

Three helpers in `marlin_utils`, wired through **every** remaining consumer of a shared per-layer workspace (FP8-Marlin dense prepare, NVFP4/MXFP4 prepares, modelopt FP8/FP4 branches, compressed-tensors w8a16_fp8 / wNa16 dense + MoE; the classic MarlinConfig method and `prepare_moe_fp8_layer_for_marlin` were removed from main after this fix was prepared, and `moe_runner/marlin.py` already allocates per-call workspaces — both intentionally untouched):

- `marlin_stream_workspaces_enabled()` — PDMux stream-group probe, exception-safe (always False without PDMux)
- `marlin_init_stream_workspaces(layer, **make_kwargs)` — opt-in at `process_weights_after_loading`; strict no-op without PDMux; preserves the MoE `max_blocks_per_sm=4` sizing
- `get_marlin_workspace_for_forward(layer)` — workspace keyed by `(device.index, stream.cuda_stream)`, lazily created; falls back to `layer.workspace` when not opted in

## Tests — scope stated honestly

- `test/registered/unit/layers/quantization/test_marlin_workspace_pdmux.py` (CI-runnable) pins the **mechanism**: stream-key isolation, non-opt-in fallback, kwargs sizing, PDMux-absent no-op, and a two-stream 24-deep-queue concurrency check (shallow submission never co-resides two launches and cannot reproduce the race) in a subprocess with a hard timeout. It is not an end-to-end `--enable-pdmux` service regression.
- `test/manual/quant/test_pdmux_mixed_deadlock_repro.py` (manual, ≥40GB GPU) is the **service-level** reproducer: with only Part 1 the mixed batch after a ≥4K prefill deadlocks (exit 2, embedded py-spy dump); with both parts it runs clean. Verified in both directions on A100-SXM4-40GB / Qwen3.8-27B-FP8.

## Validation

A100-SXM4-40GB / Qwen3.8-27B-FP8 / FP8-Marlin, at the environment documented in #37904 (which includes the Part 1 prerequisite): pre-fix mixed-batch deadlock 4/4–6/6 across six server iterations; post-fix fast gate 2/2 and the full 8-stage correctness campaign pass (hash-equal overlap matrix, prefix gate, CUDA-Graph+PDMux gate); A/B wall-time within ~10%.

---
### CI States

Latest PR Test (Base): :x: [Run #33852419865](https://github.com/sgl-project/sglang/actions/runs/33852419865)
Latest PR Test (Extra): :x: [Run #33852419546](https://github.com/sgl-project/sglang/actions/runs/33852419546)
Latest PR Test (AMD ROCm 7.2): :x: [Run #33852419916](https://github.com/sgl-project/sglang/actions/runs/33852419916)

### CODE REVIEW

**Comment by mikewong23571** [comment]
Lint failure fixed (ruff-format on the new test file). The `pr-gate` jobs report *Missing required label 'run-ci'* — per the contribution guide that label triggers the CI test matrix and needs to be added by a maintainer; happy to re-run anything else needed from my side.
